contracts/strategies/dynamic/DynamicLPStrategy.sol

```diff
@@ -0,0 +1,227 @@
+
+        subStrategies.push(subStrategy);
+        emit LogSubStrategyAdded(address(subStrategy));
+
+        if (address(currentSubStrategy) == address(0)) {
+            currentSubStrategy = subStrategy;
+
+            emit LogSubStrategyChanged(address(0), address(currentSubStrategy), 0, 0, 0, 0);
+        }
+    }
+
+    /// @param index the index of the next strategy to use
+    /// @param maxSlippageBps maximum tolerated amount of basis points of the total migrated
+    ///                   5 = 0.05%
+    ///                   10_000 = 100%
+    /// @param minDustAmount0 when the new strategy needs to wrap the token0 and token1 from previousSubStrategy
+    ///                     unwrapped token0 and token1, after initial addLiquidity, what minimum remaining
+    ///                     amount left in the contract (from new pair imbalance),
+    ///                     should be considered to swap again for more liquidity. Set to 0 to ignore.
+    /// @param minDustAmount1 same as minDustAmount0 but for token1
+    function changeStrategy(
+        uint256 index,
+        uint256 maxSlippageBps,
+        uint256 minDustAmount0,
+        uint256 minDustAmount1
+    ) public onlyExecutor {
+        require(index < subStrategies.length, "invalid index");
+
+        IDynamicSubLPStrategy previousSubStrategy = currentSubStrategy;
+        currentSubStrategy = subStrategies[index];
+        require(previousSubStrategy != currentSubStrategy, "already current");
+
+        /// @dev the next sub strategy is not using the same strategy token
+        /// and requires a convertion
+        if (previousSubStrategy.strategyTokenIn() != currentSubStrategy.strategyTokenIn()) {
+            /// @dev unwrap needs send the token0 and token1 to the next strategy directly
+            (uint256 amountFrom, uint256 priceAmountFrom) = previousSubStrategy.withdrawAndUnwrapTo(currentSubStrategy);
+
+            /// @dev wrap from the tokens sent from the previous strategy
+            (uint256 amountTo, uint256 priceAmountTo) = currentSubStrategy.wrapAndDeposit(minDustAmount0, minDustAmount1);
+
+            uint256 minToteraledPrice = priceAmountFrom - ((priceAmountFrom * maxSlippageBps) / 10_000);
+
+            require(priceAmountTo >= minToteraledPrice, "maximumBps exceeded");
+
+            emit LogSubStrategyChanged(
+                address(previousSubStrategy),
+                address(currentSubStrategy),
+                amountFrom,
+                priceAmountFrom,
+                amountTo,
+                priceAmountTo
+            );
+        }
+    }
+
+    /// @inheritdoc IStrategy
+    function skim(uint256 amount) external override onlyValidStrategy {
+        /// @dev bentobox transfers the token in this strategy so we need to
+        /// forward them to the sub strategy so that the specific skim can work.
+        ERC20(strategyToken).transfer(address(currentSubStrategy), amount);
+        currentSubStrategy.skim(amount);
```

Comment:
How can the substrategy deal with this differing token?

Reply:
if the strategy on bentobox is hooked on a joeLP token, for example, and that our current substrategy is using pangolinLP it's not possible to skim/withdraw/rebalance/exit as it would require converting the pangolinLP to joeLP before they are transfered from/to the sub strategy. There's a modifier onlyValidStrategy that checks if the current subStrategy is a "valid" one, so you need to changeStrategy to joeLP substrat first

It could be possible but it would require adding extra parameters similar to changeStrategy which I was finding "convulated" to add, like temporary parameters so that when bentobox calls them it knows the parameters, similar to safeHarvest I would say.

sub strategies, when farming, convert their rewards to the strategy token (in this case, joeLP) but when switching from strategy A to B (joeLP to pangolinLP), the convertion is done once using changeStrategy with constraints, like we discussed, to avoid slippage/frontrunning.